reconcile/utils/secret_reader.py

```diff
@@ -106,3 +112,66 @@ def read_all(self, secret: Mapping[str, str]):
                 raise SecretNotFound(*e.args) from e
 
         return data
+
+
+class TypedSecretReader(SecretReaderBase):
```

Comment:
Just had a chat with @geoberle about this. It might make sense to inject the Reader classes you mentioned (VaultSecretReader, ConfigSecretReader) into the run() function. I.e., `create_secret_reader` actually happens at `run_integration()` time.

Reply:
That was basically the motivation behind: https://gitlab.cee.redhat.com/service/dev-guidelines/-/merge_requests/140

If we use dependency injection, then we'd naturally come to the conclusion that `SecretReader` wouldn't be created in the clients themselves and instead would probably end up in `run` and used to read secrets there.

Example:

```python
# In some run() function
secret_reader = create_secret_reader(config)
some_api_token = secret_reader.read(some_secret_path)
# SomeVendorClient being the third-party client released by the vendor (ex. boto3 for AWS)
# This could also be offloaded to some create_* function
vendor_client = SomeVendorClient(some_api_token)
# SomeApiClient is the client we write as a wrapper, in utils/ (ex. AWSApi)
our_wrapper = SomeApiClient(vendor_api_client)
```
Now the client doesn't care about `settings`, `SecretReader`s, or anything else so it's not tightly coupled like what we have today. It also doesn't even need to create the underlying vendor client, so it should be easier to test overall.

Reply:
Made some changes as suggested above:

- added a `create_secret_reader(settings: Optional[SupportsVaultSettings]) -> TypedSecretReader` function. 
- `TypedSecretReader` is now the base class for `VaultSecretReader` and `ConfigSecretReader`.

@geoberle Staying backwards compatible with `SecretReader` is pretty hard, as it would require to also pass more arguments to the `__init__()` function, such as `settings: SupportsVaultSettings`. I.e., `SecretReader` would become a rather big class wrapping typed and untyped approaches + vault and config reader. For that reason I went now with @steveteahan 's suggestion. Any thoughts on this are very welcome! :bow:

Reply:
The changes look good to me! I added a few other comments and happy to discuss further as needed.

Reply:
i like the changes as well. have a look at the other comment i wrote